stellargraph/layer/graph_attention.py

```diff
@@ -0,0 +1,481 @@
+        F_in = int(input_shape[0][-1])
+
+        # Initialize weights for each attention head
+        for head in range(self.attn_heads):
+            # Layer kernel
+            kernel = self.add_weight(
+                shape=(F_in, self.F_out),
+                initializer=self.kernel_initializer,
+                regularizer=self.kernel_regularizer,
+                constraint=self.kernel_constraint,
+                name="kernel_{}".format(head),
+            )
+            self.kernels.append(kernel)
+
+            # # Layer bias
+            if self.use_bias:
+                bias = self.add_weight(
+                    shape=(self.F_out,),
+                    initializer=self.bias_initializer,
+                    regularizer=self.bias_regularizer,
+                    constraint=self.bias_constraint,
+                    name="bias_{}".format(head),
+                )
+                self.biases.append(bias)
+
+            # Attention kernels
+            attn_kernel_self = self.add_weight(
+                shape=(self.F_out, 1),
+                initializer=self.attn_kernel_initializer,
+                regularizer=self.attn_kernel_regularizer,
+                constraint=self.attn_kernel_constraint,
+                name="attn_kernel_self_{}".format(head),
+            )
+            attn_kernel_neighs = self.add_weight(
+                shape=(self.F_out, 1),
+                initializer=self.attn_kernel_initializer,
+                regularizer=self.attn_kernel_regularizer,
+                constraint=self.attn_kernel_constraint,
+                name="attn_kernel_neigh_{}".format(head),
+            )
+            self.attn_kernels.append([attn_kernel_self, attn_kernel_neighs])
+        self.built = True
+
+    def call(self, inputs, **kwargs):
+        """
+
+        Args:
+            inputs (list): list of inputs with 2 items: node features (matrix of size N x F),
+                and graph adjacency matrix (size N x N), where N is the number of nodes in the graph,
+                F is the dimensionality of node features
+            **kwargs:
+
+        Returns:
+
+        """
+        X = inputs[0]  # Node features (N x F)
+        A = inputs[1]  # Adjacency matrix (N x N)
+        # Convert A to dense tensor - needed for the mask to work
+        if K.is_sparse(A):
+            A = tf.sparse_tensor_to_dense(A, validate_indices=False)
+
+        # For the GAT model to match that in the paper, we need to ensure that the graph has self-loops,
+        # since the neighbourhood of node i in eq. (4) includes node i itself.
+        # Adding self-loops to A via setting the diagonal elements of A to 1.0:
+        # N = kwargs.get("num_nodes")
+        # get the number of nodes from inputs[1] directly, rather than passing it via kwargs
+        N = inputs[1]._keras_shape[-1]
+        if N is not None:
+            # create self-loops
+            A = tf.linalg.set_diag(A, K.cast(np.ones((N,)), dtype="float"))
```

Comment:
I've added an argument "add_self_loops" to `GAT.node_model` method, which toggles adding self-loops in the `GraphAttention` layers of the GAT model. The default value is set to `True` so that the user does not have to worry about adding self-loops to make sure the resulting GAT layers match those in the paper, yet still allowing advanced users to switch self-loops off if needed.